### PR TITLE
Add document for running JS IPFS in the browser

### PR DESCRIPTION
**Pull request by tapaswenipathak**
Fixes https://github.com/ipfs/js-ipfs/issues/2231.

### CODE REVIEW

**Comment by alanshaw** [comment]
Thanks @tapaswenipathak I'll get round to reviewing this asap!

**Comment by tapaswenipathak** [comment]
On Wed, Aug 28, 2019 at 2:04 PM Alan Shaw <notifications@github.com> wrote:
>
> Thanks @tapaswenipathak I'll get round to reviewing this asap!

Sure, thanks!

File: doc/using-js-ipfs-in-the-browser.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    // convert Buffer back to string
+    console.log(data.toString())
+  })
+})
+```
+
+JS IPFS Restriction in Web Browser
+------------------------------------------
+
+- There is no DHT in js-ipfs.
```
Comment (daviddias) [inline doc/using-js-ipfs-in-the-browser.md]: This is no longer accurate. There is a DHT just not enabled by default. This PR is working on integrating it by default https://github.com/ipfs/js-ipfs/pull/1994

File: doc/using-js-ipfs-in-the-browser.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+- Unable in opening TCP connections even if one sets up delegating routing the 
+issue will persist for connecting most of peers in the swarm as described a
+little above.
+
+- Connection with websocket ports should be adhering with [secure contexts](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts) limitations.
+
+JS IPFS Best Practices
+----------------------
+
+- Run in fedrated model.
```
Comment (daviddias) [inline doc/using-js-ipfs-in-the-browser.md]: @tapaswenipathak what do you mean by this?

Comment (tapaswenipathak) [reply doc/using-js-ipfs-in-the-browser.md]: I don't remember now.

**Comment by tapaswenipathak** [comment]
@daviddias: you can review now.

File: doc/using-js-ipfs-in-the-browser.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    // convert Buffer back to string
+    console.log(data.toString())
+  })
+})
+```
+
+JS IPFS Restriction in Web Browser
+------------------------------------------
+
+- DHT in js-ipfs is not by default.
```
Comment (daviddias) [inline doc/using-js-ipfs-in-the-browser.md]:
```suggestion
- DHT in js-ipfs is not enabled by default.
```

File: doc/using-js-ipfs-in-the-browser.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+JS IPFS Best Practices
+----------------------
+
+- Configure nodes for using webrtc or websocket-star.
+- Run independant instance as default one are under high load and YMMV.
+- Make sure content you are implementing is cached in go-ipfs
+  - manually `pin` or preload CIDs of interest with `refs -r` beforehand.
+  - preload content on the fly using [preload](https://github.com/ipfs/js-ipfs#optionspreload) or
+    configuring [delgated routing](https://github.com/ipfs/js-ipfs#configuring-delegate-routers).
```
Comment (daviddias) [inline doc/using-js-ipfs-in-the-browser.md]:
```suggestion
    configuring [delegated routing](https://github.com/ipfs/js-ipfs#configuring-delegate-routers).
```

File: doc/using-js-ipfs-in-the-browser.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+JS IPFS Restriction in Web Browser
+------------------------------------------
+
+- DHT in js-ipfs is not enabled by default.
+
+  In browser a JS IPFS node is more restricted because of the same origin policy
+  that decrease the discover/connect with peers, external peers without rendezvous
+  and relay servers, delegating peer/content routing and preload servers for content
+  discoverablilty hard and erroneous even with swarm peers.
```
Comment (lidel) [inline doc/using-js-ipfs-in-the-browser.md]: Let's rephrase this section to focus on discovery and connectivity, something like:

> This means js-ipfs running in the browser is limited to Web APIs available on a web page.
> There is no raw TCP sockets nor low level UDP, only WebRTC and Websockets.
> Discovery and connectivity to other peers is achieved with a mix of rendezvous and relay servers,
> delegated peer/content routing and preload servers.

File: doc/using-js-ipfs-in-the-browser.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+- If the browser lacks TCP, you can set up delegating routing. The issue might still
+  persist for connecting most of peers in the swarm as described above.
+
+- Connection with websocket ports should be adhering with [secure contexts](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts) limitations.
+
+JS IPFS Best Practices
+----------------------
+
+- Configure nodes for using webrtc or websocket-star.
+- Run independant instance as default one are under high load and YMMV.
```
Comment (lidel) [inline doc/using-js-ipfs-in-the-browser.md]:
```suggestion
- Run your own instance of `*-star` signaling service. The default ones are under high load and YMMV.
```

File: doc/using-js-ipfs-in-the-browser.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+JS IPFS is the implementation of IPFS protocol in JavaScript. It can run on any
+evergreen browser, inside a service or web worker, as a web extention and in Node.js.
+
+The documentation details more information about running JS IPFS in the browser and
+boosting the speed with reducing the errors you face often because of limitation of
+js-ipfs.
+
+There are few different components for implementing JS IPFS in the browser. These
+are webrtc and websocket-star which will make it easier for implemented JS IPFS
```
[the remark this answers is not in the document]
Comment (tapaswenipathak) [reply doc/using-js-ipfs-in-the-browser.md]: Having grammar fixes: 873d147a8eae196cfeb67ba1f04a492758bbe534.

File: doc/using-js-ipfs-in-the-browser.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    // convert Buffer back to string
+    console.log(data.toString())
+  })
+})
+```
+
+JS IPFS Restriction in Web Browser
+------------------------------------------
+
+- DHT in js-ipfs is not enabled by default.
```
[the remark this answers is not in the document]
Comment (tapaswenipathak) [reply doc/using-js-ipfs-in-the-browser.md]: ref: https://github.com/ipfs/js-ipfs/pull/2392#discussion_r333407725.

**Comment by tapaswenipathak** [comment]
@lidel: Very creative statements/writeups.

**Comment by autonome** [comment]
This is a great document! It answers many questions in a concise way.

Any reason to not merge this as it is, and continue to iterate from there?

**Pull request merged**